Add `ignore_codingkeys` parameter to nesting

braker1nine opened the pull request:
Resolves #5641 

This very explicitly targets `enum`'s named `CodingKeys` which conform to `CodingKey`. To provide an exception for `Codable` users who have to provide custom keys. 


OK ✅ 
```swift
struct Outer {
    struct Inner: Codable {
        let identifier: Int
        enum CodingKeys: String, CodingKey {
            case identifier = "id"
        }
    }
}
```
---

Not OK ❌ 
```swift
struct Outer {
    struct Inner {
        let identifier: Int
        enum SomeKeys: String, CodingKey {
            case identifier = "id"
        }
    }
}
```

SimplyDanny commented on Source/SwiftLintCore/Extensions/SwiftSyntax+SwiftLint.swift:
```diff
@@ -184,6 +184,25 @@ public extension EnumDeclSyntax {
         }
     }
+
+    /// True if this enum is a `CodingKey`
```
```suggestion
    /// True if this enum is a `CodingKey`. For that, it has to be named `CodingKeys` and must conform to the `CodingKey` protocol. 
```

SimplyDanny commented on Source/SwiftLintCore/Extensions/SwiftSyntax+SwiftLint.swift:
```diff
@@ -184,6 +184,25 @@ public extension EnumDeclSyntax {
     }
+
+    /// True if this enum is a `CodingKey`
+    var isCodingKey: Bool {
```
The enum itself is not a coding key, but it defines the coding keys, doesn't it? So what about the name `definesCodingKeys` instead?

SimplyDanny commented on Source/SwiftLintCore/Extensions/SwiftSyntax+SwiftLint.swift:
```diff
@@ -184,6 +184,25 @@ public extension EnumDeclSyntax {
+            return false
+        }
+        
+        guard self.name.text == "CodingKeys" else {
```
```suggestion
        guard name.text == "CodingKeys" else {
```

Can also be added to the previous `guard`.

SimplyDanny commented on Source/SwiftLintCore/Extensions/SwiftSyntax+SwiftLint.swift:
```diff
@@ -184,6 +184,25 @@ public extension EnumDeclSyntax {
+        }
+
+        return inheritedTypeCollection.contains { element in
+            guard let identifier = element.type.as(IdentifierTypeSyntax.self)?.name.text else {
+                return false
+            }
+
+            return identifier == "CodingKey"
```
```suggestion
            element.type.as(IdentifierTypeSyntax.self)?.name.text == "CodingKey"
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Metrics/NestingRule.swift:
```diff
@@ -64,7 +64,9 @@ private extension NestingRule {
 
         override func visit(_ node: EnumDeclSyntax) -> SyntaxVisitorContinueKind {
-            validate(forFunction: false, triggeringToken: node.enumKeyword)
+            if !(configuration.ignoreCodingKeys && node.isCodingKey) {
```
```suggestion
            if !configuration.ignoreCodingKeys || !node.isCodingKey {
```

SimplyDanny commented on Tests/SwiftLintFrameworkTests/NestingRuleTests.swift:
```diff
@@ -551,4 +551,38 @@ final class NestingRuleTests: SwiftLintTestCase {
+            .init("""
+                struct Outer {
+                    struct Inner {
+                        enum Example: String, CodingKey {
```
```suggestion
                        ↓enum Example: String, CodingKey {
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/RuleConfigurations/NestingConfiguration.swift:
```diff
@@ -15,6 +15,8 @@ struct NestingConfiguration: RuleConfiguration {
     private(set) var alwaysAllowOneTypeInFunctions = false
     @ConfigurationElement(key: "ignore_typealiases_and_associatedtypes")
     private(set) var ignoreTypealiasesAndAssociatedtypes = false
+    @ConfigurationElement(key: "ignore_codingkeys")
```
```suggestion
    @ConfigurationElement(key: "ignore_coding_keys")
```

SimplyDanny reviewed (changes requested):
Thank you for the contribution! Looks good in principle.

What about the following example?

```swift
struct Outer {
  enum CodingKeys: String, CodingKey {
    case id
    
    struct S {}
  }
}
```

I think, the rule should trigger on `S`.

You need to run `IntegrationTests` locally and fix all linting issues it reports and update the configuration set reference file.

SimplyDanny commented on CHANGELOG.md:
```diff
@@ -16,6 +16,10 @@
   improvements done in the [SwiftSyntax](https://github.com/apple/swift-syntax)
   library.
 
+* Add `ignore_codingkeys` parameter to `nesting` rule.  
```
Please explain briefly what this option does.

```suggestion
* Add `ignore_coding_keys` parameter to `nesting` rule.  
```

braker1nine commented:
FYI I'm out of town until next weekend. I'll jump back into this once I'm back

braker1nine commented:
> FYI I'm out of town until next weekend. I'll jump back into this once I'm back

It wasn't a year long vacation but here we are a year later 😓 

 
> What about the following example?
> 
> ```swift
> struct Outer {
>   enum CodingKeys: String, CodingKey {
>     case id
>     
>     struct S {}
>   }
> }
> ```
> 
> I think, the rule should trigger on `S`.

I've updated this so that this will now trigger the rule.

> You need to run `IntegrationTests` locally and fix all linting issues it reports and update the configuration set reference file.

I think I've got this done? I moved the `EnumDeclSyntax` extension into its own file because the `SwiftSyntax+SwiftLint` file was too long and failing linting.

I don't fully understand what is failing now?

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Metrics/NestingRule.swift:
```diff
@@ -152,6 +152,12 @@ private extension NestingRule {
             }
+
+            // if current defines coding keys and we're ignoring coding keys, then skip nesting rule
+            if configuration.ignoreCodingKeys && definesCodingKeys {
```
Can we move this into `visit(_ node: EnumDeclSyntax)` to avoid the extra argument?

braker1nine replied:
Hmmm yeah I think I could I would need to ensure we push another item onto `self.levels` in `visit(_ node: EnumDeclSyntax)`. To ensure the `CodingKeys` add another level to the nesting count without violating it themselves. 

```swift
override func visit(_ node: EnumDeclSyntax) -> SyntaxVisitorContinueKind {

    // if current defines coding keys and we're ignoring coding keys, then skip nesting rule
    // push another level on and proceed to visit children
    if configuration.ignoreCodingKeys && node.definesCodingKeys {
        levels.push(levels.lastIsFunction)
    } else {
        validate(forFunction: false, triggeringToken: node.enumKeyword)
    }

    return .visitChildren
}
```

But I that does seem a little simpler? Doesn't require all the other `visit` methods to add this parameter

SimplyDanny commented on Tests/BuiltInRulesTests/NestingRuleTests.swift:
```diff
@@ -555,4 +555,48 @@ final class NestingRuleTests: SwiftLintTestCase {
         verifyRule(description, ruleConfiguration: ["ignore_typealiases_and_associatedtypes": true])
     }
+
+    func testNestingWithoutCodingKeys() {
```
You can move these examples directly into `NestingRuleExamples`. Any options deviating from the default will be added to the online documentation, so it's clear why an example triggers or why it doesn't.

SimplyDanny replied:
In the documentation that includes these examples later, we now also show the configuration in case it deviates from the default one. So having them as normal examples is even nicer as people also see how options influence the rule's behavior.

SimplyDanny replied:
The feature that configurations are added to the examples in the documentation is quite new and we won't convert existing tests. But we should prefer this way of documenting examples from now on.

SimplyDanny commented on Tests/BuiltInRulesTests/NestingRuleTests.swift:
```diff
@@ -555,4 +555,48 @@ final class NestingRuleTests: SwiftLintTestCase {
+                  }
+                }
+            """),
+        ])
```
Let's have another example that triggers on a `CodingKeys` enum when the option is set to `false` (the default).

SimplyDanny reviewed (changes requested):
Glad you are back! 🥳 I have a few more comments ...

SimplyDanny commented:
For the build, please rebase onto the current `main`. The results are outdated and there is conflict as well.

braker1nine commented on Source/SwiftLintCore/Extensions/SwiftSyntax+SwiftLint.swift:
```diff
@@ -169,30 +169,6 @@ public extension VariableDeclSyntax {
     }
 }
 
-public extension EnumDeclSyntax {
```
👍 is there a preferred place to put that disable? Top of the source file?

SimplyDanny replied:
Yes, typically below imports.

SimplyDanny commented on CHANGELOG.md:
```diff
@@ -63,6 +63,12 @@
   `max_number_of_single_line_parameters` and detect mixed formatting.  
   [GandaLF2006](https://github.com/GandaLF2006)
 
+* Add `ignore_coding_keys` parameter to `nesting` rule.  
+* Add `ignore_coding_keys` parameter to `nesting` rule. Setting this to true prevents
```
```suggestion
* Add `ignore_coding_keys` parameter to `nesting` rule. Setting this to true prevents
```

SimplyDanny reviewed (approved):
Looks good now. Thanks for the updates, @braker1nine!
